src/Common/tests/System/Net/Http/Http2LoopbackServer.cs

```diff
@@ -0,0 +1,160 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Security;
+using System.Security.Cryptography.X509Certificates;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Test.Common
+{
+    public class Http2LoopbackServer : IDisposable
+    {
+        private Socket _listenSocket;
+        private Stream _connectionStream;
+        private Http2Options _options;
+        private Uri _uri;
+
+        public Http2LoopbackServer(Http2Options options)
+        {
+            _options = options;
+        }
+
+        public Uri CreateServer()
+        {
+            _listenSocket = new Socket(_options.Address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            _listenSocket.Bind(new IPEndPoint(_options.Address, 0));
+            _listenSocket.Listen(_options.ListenBacklog);
+
+            var localEndPoint = (IPEndPoint)_listenSocket.LocalEndPoint;
+            string host = _options.Address.AddressFamily == AddressFamily.InterNetworkV6 ?
+                $"[{localEndPoint.Address}]" :
+                localEndPoint.Address.ToString();
+
+            string scheme = _options.UseSsl ? "https" : "http";
+
+            _uri = new Uri($"{scheme}://{host}:{localEndPoint.Port}/");
+
+            return _uri;
+        }
+
+        public async Task SendConnectionPrefaceAsync()
+        {
+            Frame emptySettings = new Frame(0, FrameType.Settings, FrameFlags.None, 0);
+            await WriteFrameAsync(emptySettings).ConfigureAwait(false);
+
+            // We also need to ack the client settings header sent with the preface.
+            // We could piggyback along with another frame, but this is simpler.
+            Frame ack = new Frame(0, FrameType.Settings, FrameFlags.Ack, 0);
+            await WriteFrameAsync(emptySettings).ConfigureAwait(false);
+        }
+
+        public async Task WriteFrameAsync(Frame frame)
+        {
+            byte[] writeBuffer = new byte[Frame.FrameHeaderLength + frame.Length];
+            frame.WriteTo(writeBuffer);
+            await _connectionStream.WriteAsync(writeBuffer, 0, writeBuffer.Length).ConfigureAwait(false);
+        }
+
+        public async Task<Frame> ReadFrameAsync(TimeSpan timeout)
```

Comment:
Calling this without a timeout can cause the tests to hang indefinitely if no frame is received. Since hangs can be difficult to diagnose I decided to make a timeout mandatory. In the tests I used an arbitrary value of 30 seconds. I can rework this if anyone thinks there's a better way.